### ISSUE TITLE
The device X owned by Y does not have a bundle and will be skipped. The owner should fix this, or remove the device from the list

### ISSUE THREAD

**Issue by dreamflasher** [gkdr/lurch#106]
We have the following setup:
Person 1 (me), Client A: Conversations 
Person 1 (me), Client B: Pidgin, Windows
Person 2 (they), Client A: Pidgin
Person 2 (they), Client B: Conversations

Messages between 1A and 2A work completely fine. Messages between 1B and the rest don't work at all. 

Incoming messages: `I sent you an OMEMO encrypted message but your client doesn’t seem to support that. Find more information on https://conversations.im/omemo`
```<message to='person1/285227426255934208143014157' from='person1' type='chat'>
	<sent xmlns='urn:xmpp:carbons:2'>
		<forwarded xmlns='urn:xmpp:forward:0'>
			<message lang='en' to='person2' from='person1/Conversations.yCF7' type='chat' id='70ddb166-6481-44f0-a868-5e2792baab23'>
				<archived xmlns='urn:xmpp:mam:tmp' by='person1' id='1526721492613074'/>
				<stanza-id xmlns='urn:xmpp:sid:0' by='person1' id='1526721492613074'/>
				<encrypted xmlns='eu.siacs.conversations.axolotl'>
					<header sid='1318061081'>
						<key rid='1103906935'>MwohBTU/HZOMOyKsAJZ6SiP5a1PJ3wW2vHCQMxEGSb5m7ZJREEAYBSIwJquSnfBwxJoF9PHVm50bdUdOzfsiij9luqyW03oNhvYb4cY21qjQZmpHHR4cAODDeNc+voNhyd8=</key>
						<key rid='1488963539'>MwohBUMf3/CJmjdxWSO+5ZREWMjSkLvlXmZTQyhz8YoypwhhEOEBGAAiMEvzqqelGGhRaZjC4lpKeIIn9OCdMcdlMPPGTkGg7Ddimv0+qBaQIuNBNTqMuttS7L9jyF1H5yBm</key>
						<key rid='1558494486' prekey='true'>MwgPEiEF+D7cmk/Z5f+lznqFTVtz+4RR8XAYavhAmqC605HJylQaIQWfbvpNX/UmkE+XI0lSq+ZniIbhtdZuQq5wTNIg/m5qXiJiMwohBbVwMw+iinYNmUGVruml2We9b6NqlTGYGi7uvZMdxeRKEEcYACIwCS1kE1cbIBKq9CPcd8gchak6qfXJ8PYUbJ1oDCDCBgxPD3buyjJJHUUqoYucYKk+f40Se2A5yQoomYjA9AQwAA==</key>
						<key rid='1810888295'>MwohBcMagQtFqGVuZGkb3HU5upN2NZn+dqvygIjdklsBlaYzEAMYACIwtFwtILWKeqasjPoxw7tvcPr/46gl8Yja4jbjdAMN22vAbWG7oU1oXmsMsfxid2piaJFLSqA1hO4=</key>
						<iv>h34rgAiqpTZliKpZC1f2xQ==</iv>
					</header>
					<payload>ovghBw==</payload>
				</encrypted>
				<markable xmlns='urn:xmpp:chat-markers:0'/>
				<origin-id xmlns='urn:xmpp:sid:0' id='70ddb166-6481-44f0-a868-5e2792baab23'/>
				<store xmlns='urn:xmpp:hints'/>
				<encryption xmlns='urn:xmpp:eme:0' namespace='eu.siacs.conversations.axolotl' name='OMEMO'/>
				<delay xmlns='urn:xmpp:delay' stamp='2018-05-19T09:18:07.731Z'/>
				<body>I sent you an OMEMO encrypted message but your client doesn’t seem to support that. Find more information on https://conversations.im/omemo</body>
			</message>
		</forwarded>
	</sent>
</message>
<message lang='en' to='person1/Conversations.yCF7' from='person2/Conversations.5SKj' type='chat'>
	<archived xmlns='urn:xmpp:mam:tmp' by='person1' id='1526721493398843'/>
	<stanza-id xmlns='urn:xmpp:sid:0' by='person1' id='1526721493398843'/>
	<received xmlns='urn:xmpp:chat-markers:0' id='70ddb166-6481-44f0-a868-5e2792baab23'/>
	<store xmlns='urn:xmpp:hints'/>
</message>
```

Outgoing messages: `The device A owned by 1 does not have a bundle and will be skipped. The owner should fix this, or remove the device from the list.

```
<iq lang='en' to='person1/285227426255934208143014157' from='person1' type='error' id='person1#2147483647#2048213375'>
	<pubsub xmlns='http://jabber.org/protocol/pubsub'>
		<items node='eu.siacs.conversations.axolotl.bundles:2147483647' max_items='1'/>
	</pubsub>
	<error code='404' type='cancel'>
		<item-not-found xmlns='urn:ietf:params:xml:ns:xmpp-stanzas'/>
		<text xmlns='urn:ietf:params:xml:ns:xmpp-stanzas' lang='en'>Node not found</text>
	</error>
</iq>
<message type='chat' id='purplea272ae45' to='person2'>
	<active xmlns='http://jabber.org/protocol/chatstates'/>
	<encrypted xmlns='eu.siacs.conversations.axolotl'>
		<header sid='1558494486'>
			<key rid='1318061081'>Mwh+EiEFXu2uo5OtXy6swTzreyeKWcLU05GZd+YJmwwrtGO3iW8aIQU1W9EaGWe5Ma+2E5kk9EKCBOsHQkELbqiSsWAuv98PIiJiMwohBQjxpI+z5EUKjuvQ0L8AwdxQaZ6rUnwD1xmPK9/EMwJOEAMYACIw+4/zcOgZ2VHaXVyAxjqG3uhN6O6xvfW72o66g8Cx4fh+mkeSbxRydwpdEcHEikQbbO/9PyWeUA8olvqS5wUwAg==</key>
			<key rid='1488963539'>MwjQARIhBbC//WbibvGFAfj/VVdx8boNpuVcqEI30QPZZF42yotCGiEFOVvRGhlnuTGvthOZJPRCggTrB0JBC26okrFgLr/fDyIiYjMKIQUrtawjd3ZFlPx3+5PDDhCYnROjulJaPgpSPxuJfPSjYxADGAAiMFlf7Ktj5unV8lrvyxsyVYJZ+9t50XKgP9VNBjV7sZuQFVUq9tznQxHkVXh5ORNFiTAgKLxPV89vKJb6kucFMAI=</key>
			<key rid='1103906935'>MwgwEiEFwdFkze+GHNb4z+/K5xDBCBUksDTPE5IkMZse4DPMYkgaIQU5W9EaGWe5Ma+2E5kk9EKCBOsHQkELbqiSsWAuv98PIiJiMwoh1Q0noz5rvrlObWObi+J8xJ8bI5UgC2h+q3kE672K/rc/EAEYACIw0pJ2xfi3oOEvUbtxXC1Xh7Zu/H407lpjH5ONZZH0q50rHPZlgSSVQCJrrylPU5U0MK1OqPwT1VMolvqS5wUwAA==</key>
			<key rid='1810888295'>MwgMEiEFgiUMl451/teXOgwPVkB6D7ZqJW7JjZ7p+DCc4a8CBXYaIQU5W9EaGWe5Ma+2E5kk9EKCBOsHQkELbqiSsWAuv98PIiJiMwoh3bk7zrSa+fWGt7mU4JIigvqg+klOY4KVpn5K5V5nmJREEAEYACIw8+FhmRopBzMsH2BBScnjksxss0tSt2ESlVkjTk3u2EoKk2fQdRVYucCOt+dsNZzgF0FzrByePboolvqS5wUwAQ==</key>
			<iv>MFxVhP3x0JiH3+dN0fvfSQ==</iv>
		</header>
		<payload>NulIWw==</payload>
	</encrypted>
	<encryption xmlns='urn:xmpp:eme:0' namespace='eu.siacs.conversations.axolotl' name='OMEMO'/>
	<store xmlns='urn:xmpp:hints'/>
</message>
<message to='person1/285227426255934208143014157' from='person1' type='headline'>
	<event xmlns='http://jabber.org/protocol/pubsub#event'>
		<items node='eu.siacs.conversations.axolotl.bundles:1488963539'>
			<item id='5F6B012742CE6'>
				<bundle xmlns='eu.siacs.conversations.axolotl'>
					<signedPreKeyPublic signedPreKeyId='2'>BeHGpi24CH9081/9X8p10pb8mtS1ymi+Ra6F3xnvuURm
</signedPreKeyPublic>
					<signedPreKeySignature>pNILZh2LMBf/1hI4xaMWiVXqBqb78c8nlr55lxSBMk9Sca6kZ6rVnMbkvZ4FAJqDW6TTHgcs3xwk
jQpWrD6fDw==
</signedPreKeySignature>
					<identityKey>BXq3CfR7NqiLbEKcGVtVhFqgFJxFroR15xtmbT560Fds
</identityKey>
					<prekeys>
						<preKeyPublic preKeyId='186'>BQ39qJ3KDLKvWxtWu7LCP1byLRZy/Zx5AfcY2gHe/T5H
</preKeyPublic>
... many prekeys
					</prekeys>
				</bundle>
			</item>
		</items>
	</event>
	<addresses xmlns='http://jabber.org/protocol/address'>
		<address jid='person1/phone' type='replyto'/>
	</addresses>
</message>
```

The other log:

```(11:28:01) lurch: lurch_msg_finalize_encryption: person1 has device without session 2147483647, requesting bundle
(11:28:01) lurch: lurch_bundle_request_do: person1/ is requesting bundle from person1:2147483647
(11:28:01) jabber: Sending (ssl) (person1/534200512930957414543020309): <iq type='get' to='person1' id='person1#2147483647#695693397'><pubsub xmlns='http://jabber.org/protocol/pubsub'><items node='eu.siacs.conversations.axolotl.bundles:2147483647' max_items='1'/></pubsub></iq>
(11:28:01) lurch: lurch_bundle_request_do: ...request sent
(11:28:01) g_log: xmlnode_to_str_helper: assertion `node != NULL' failed
(11:28:01) g_log: jabber_send_raw: assertion `data != NULL' failed
(11:28:01) jabber: Sending (ssl) (person1/534200512930957414543020309): <message type='chat' id='purple6b8f35cf' to='person2'><active xmlns='http://jabber.org/protocol/chatstates'/></message>
(11:28:01) jabber: Recv (ssl)(495): <iq xml:lang='en' to='person1/534200512930957414543020309' from='person1' type='error' id='person1#2147483647#695693397'><pubsub xmlns='http://jabber.org/protocol/pubsub'><items node='eu.siacs.conversations.axolotl.bundles:2147483647' max_items='1'/></pubsub><error code='404' type='cancel'><item-not-found xmlns='urn:ietf:params:xml:ns:xmpp-stanzas'/><text xml:lang='en' xmlns='urn:ietf:params:xml:ns:xmpp-stanzas'>Node not found</text></error></iq>
(11:28:01) lurch: lurch_bundle_request_cb: person1 received bundle update from person1:2147483647
(11:28:01) lurch: lurch_msg_encrypt_for_addrs: trying to encrypt key for 5 devices
(11:28:01) lurch: lurch_key_encrypt: encrypting key for person2:1810888295
(11:28:01) lurch: lurch_key_encrypt: encrypting key for person2:1103906935
(11:28:01) lurch: lurch_key_encrypt: encrypting key for person1:1488963539
(11:28:01) lurch: lurch_key_encrypt: encrypting key for person1:1318061081```

**Comment by DreamFlasher** [comment]
I verified again, and actually messages from 1B to 2 work, it's just that my own devices don't receive each other's messages. And throw an error with every message.

The latest versions of the follwing software are installed: Pidgin, Carbons, Lurch, Conversations, and my server is fully omemo compatible (jabber.at).

There are more devices, but I wanted to focus on the problem at hand. There's also a Miranda NG client on my side, which has the same problem. My clients don't receive messages amongst each other. See https://github.com/miranda-ng/miranda-ng/issues/1255#issuecomment-390143505

**Comment by gkdr** [comment]
The problem is that you cannot build a session, because there is no bundle.

You can verify by copypasting the same line your client sent to the Pidgin XMPP Console (you might need to activate this plugin first): `<iq type='get' to='person1' id='person1#2147483647#695693397'><pubsub xmlns='http://jabber.org/protocol/pubsub'><items node='eu.siacs.conversations.axolotl.bundles:2147483647' max_items='1'/></pubsub></iq>`

The result will probably be also this line: `<iq xml:lang='en' to='person1/534200512930957414543020309' from='person1' type='error' id='person1#2147483647#695693397'><pubsub xmlns='http://jabber.org/protocol/pubsub'><items node='eu.siacs.conversations.axolotl.bundles:2147483647' max_items='1'/></pubsub><error code='404' type='cancel'><item-not-found xmlns='urn:ietf:params:xml:ns:xmpp-stanzas'/><text xml:lang='en' xmlns='urn:ietf:params:xml:ns:xmpp-stanzas'>Node not found</text></error></iq>`.

(Both are copied from your debug log.)

If there is no sudden mixup of this client inserting the wrong device ID into the bundle request (i.e. `eu.siacs.conversations.axolotl.bundles:2147483647`), or it is malformed in some other way, there is no issue with this client, as there is nothing it can do without a bundle.
PEP nodes are not persistent, maybe your XMPP server was restarted and your Conversations client did not recognize yet that its bundle is missing.

tl;dr: Please verify manually that the devices you are told don't have a bundle actually do have a bundle. Only then it is a bug of this client.

**Comment by DreamFlasher** [comment]
Thank you! Although I don't get what I can do about it now. What is a bundle? And how can I verify in Pidgin that my client has a bundle? And how can I do this in conversations (in case you know)?

**Comment by gkdr** [comment]
>What is a bundle? 

The bundle contains all the public keys required for the cryptographic handshake, i.e. "session establishment".

>And how can I verify in Pidgin that my client has a bundle?

It should be pushed every time you connect. You can confirm this by leaving the debug window open when exiting Pidgin, as it will be opened up automatically the next time you start Pidgin. Disconnecting and reconnecting the account will probably also work.

>And how can I do this in conversations (in case you know)?

I don't know how Conversations deals with this issue, but I'm sure it does somehow so by the time you read this it might already be gone.

But don't get me wrong, I'm not saying that it's not this plugin's fault just yet. If the problem persists, please open up the XMPP console in Pidgin and paste the bundle query I pointed out in there (setting the right JID and device ID).

**Comment by DreamFlasher** [comment]
Figured out part of the issue: I was using the portable version of Pidgin, and it doesn't work with that one. When using the normal installable version it works as expected.

**Comment by shtrom** [comment]
Same issue with Portable Pidgin after a reinstall, but it did work before. I'll dig up, as I think I still have the old binaries around somewhere. Thanks @DreamFlasher for pointing me in the right direction!

**Comment by shtrom** [comment]
Ok, my old lurch binaries (from @EionRobb) don't work anymore.

Playing with dependency walker, there are a few missing DLLs for libgcrypt-20.dll:
* API-MS-WIN-CORE-KERNEL32-PRIVATE-L1-1-1.DLL
* API-MS-WIN-CORE-PRIVATEPROFILE-L1-1-1.DLL
* API-MS-WIN-SERVICE-PRIVATE-L1-1-1.DLL
I'm wondering whether those missing libs could lead to unresolved symbols and, ultimately, a `null` rather than the value we want to send.

That's using an x86 Portable Pidgin (I don't think there's an x64 version) on a 64-bit Windows 8.

**Comment by shtrom** [comment]
Here's a screenshot of DependencyWalker once I configured the paths to include the various directories containing DLLs in the PortablePidgin's directory.

![lurch_deps](https://user-images.githubusercontent.com/160280/44915862-88a49100-ad77-11e8-8eb8-ad8780a266c0.PNG)

**Comment by gkdr** [comment]
Closing this since it's been a while, and it also seems to be some issue with Pidgin Portable. If this still is a problem, please open a new issue.
